### DISCUSSION TITLE
How can I allow a leading underscore in the route segment to avoid name conflict?

### DISCUSSION THREAD

**Discussion by gerardnico** [vercel/next.js#50629] [App Router]
Hello,

I'm writing an app where by default, we show the documentation pages.
It means that if there is a documentation page called `app.md` , we will show it for the request `http://example.com/app` thanks to a [catch-all route](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#catch-all-segments).

We have an app and we want to avoid name conflict by adding an underscore. `http://example.com/_app`
I have therefore created a subdirectory in the app directory named `_app` and a `page.js` file.
But it seems to not work at all as I get a 404, not found.

I couldn't find anything:
- on the [route segment name convention](https://nextjs.org/docs/app/building-your-application/routing#route-segments)
- or how I can remap the `_app` URL path to a directory. The [middelware rewrite example](https://nextjs.org/docs/pages/building-your-application/routing/middleware#conditional-statements) seems to use the URL again.
- in the [RFC](https://nextjs.org/blog/layouts-rfc#defining-routes)

Can a NextJs expert send me a pointer?

Thanks!

**Comment by gerardnico** [comment]
I just found out the configuration [skipMiddlewareUrlNormalize](https://nextjs.org/docs/app/building-your-application/routing/middleware#advanced-middleware-flags) but it has no effect.

**Comment by gerardnico** [comment]
Note that if I use two leading underscore, I'm redirected into the `catch all` route. Funny, not into the directory `__app`

**Comment by gerardnico** [comment]
I could get it working with a [middelware](https://nextjs.org/docs/app/building-your-application/routing/middleware#conditional-statements) by redirecting the URL to a directory with a trailing underscore.

```js
if (pathname === '/_app') {
        return NextResponse.rewrite(new URL('/app_', req.url));
}
```

I'm not sure if that's the way to go.
Is there any description of the name convention for a segment.
Thanks

**Comment by gerardnico** [comment]
Extra information: The Route Segment is used in the export name of a [nested layout](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#nesting-layouts).

For a `dashboard` segment, the exported layout name is `DashboardLayout` ...
```js
export default function DashboardLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return <section>{children}</section>;
}
```

Therefore, you need to follow the [naming convention of Javascript](https://262.ecma-international.org/5.1/#sec-7.6)

**Comment by jlubawy** [comment]
I ran into the same issue, but it appears you can add a prefix of `%5F` and it will serve it with a leading `_` instead 🙏 (see https://github.com/vercel/next.js/pull/48792/files#diff-afcb7c97bf71ebbcf9ecac20d9e21c1870c558de6316f800c269d64b31f5d5dbR702).
